src/append.jl

```diff
@@ -0,0 +1,168 @@
+or the `JULIA_NUM_THREADS` environment variable is set).
+
+Supported keyword arguments to `Arrow.append` include:
+  * `compress`: possible values include `:lz4`, `:zstd`, or your own initialized `LZ4FrameCompressor` or `ZstdCompressor` objects; will cause all buffers in each record batch to use the respective compression encoding
+  * `alignment::Int=8`: specify the number of bytes to align buffers to when written in messages; strongly recommended to only use alignment values of 8 or 64 for modern memory cache line optimization
+  * `dictencode::Bool=false`: whether all columns should use dictionary encoding when being written; to dict encode specific columns, wrap the column/array in `Arrow.DictEncode(col)`
+  * `dictencodenested::Bool=false`: whether nested data type columns should also dict encode nested arrays/buffers; other language implementations [may not support this](https://arrow.apache.org/docs/status.html)
+  * `denseunions::Bool=true`: whether Julia `Vector{<:Union}` arrays should be written using the dense union layout; passing `false` will result in the sparse union layout
+  * `largelists::Bool=false`: causes list column types to be written with Int64 offset arrays; mainly for testing purposes; by default, Int64 offsets will be used only if needed
+  * `maxdepth::Int=$DEFAULT_MAX_DEPTH`: deepest allowed nested serialization level; this is provided by default to prevent accidental infinite recursion with mutually recursive data structures
+  * `ntasks::Int`: number of concurrent threaded tasks to allow while writing input partitions out as arrow record batches; default is no limit; to disable multithreaded writing, pass `ntasks=1`
+"""
+function append end
+
+append(file::String; kw...) = x -> append(file::String, x; kw...)
+
+function append(file::String, tbl;
+        largelists::Bool=false,
+        compress::Union{Nothing, Symbol, LZ4FrameCompressor, ZstdCompressor}=nothing,
+        denseunions::Bool=true,
+        dictencode::Bool=false,
+        dictencodenested::Bool=false,
+        alignment::Int=8,
+        maxdepth::Int=DEFAULT_MAX_DEPTH,
+        ntasks=Inf)
+    if ntasks < 1
+        throw(ArgumentError("ntasks keyword argument must be > 0; pass `ntasks=1` to disable multithreaded writing"))
+    end
+    if compress === :lz4
+        compress = LZ4_FRAME_COMPRESSOR
+    elseif compress === :zstd
+        compress = ZSTD_COMPRESSOR
+    elseif compress isa Symbol
+        throw(ArgumentError("unsupported compress keyword argument value: $compress. Valid values include `:lz4` or `:zstd`"))
+    end
+
+    open(file, "r+") do io
+        append(io, tbl, largelists, compress, denseunions, dictencode, dictencodenested, alignment, maxdepth, ntasks)
+    end
+
+    return file
+end
+
+function append(io, source, largelists, compress, denseunions, dictencode, dictencodenested, alignment, maxdepth, ntasks)
+    bytes = Mmap.mmap(io)
+    stream = Arrow.Stream(bytes)
+    arrow_schema = Tables.schema(Tables.columns(stream))
```

Comment:
If this works, then it's probably by accident. By design, you're supposed to call `Tables.columns` on each _iterated element_ of an `Arrow.Stream`. We can, however, access the arrow schema by calling `stream.schema`. That gives us the arrow schema as a flatbuffer object. I think that's the right thing we need to validate against. For example, if we start iterating input `Tables.partitions(source)`, then we call `cols = Tables.columns(part); sch = Tables.schema(cols)`, then we can call `makeschemamsg(sch, cols)` which gives us back a flatbuffer of the schema. Then we should flip that flatbuffer around to read it `arrowsch = Meta.Schema(buf, 1)` (hopefully that works). Then we can compare `stream.schema` vs `arrowsch`.

Reply:
The alternative approach would be to iterate the first table of the `Arrow.Stream`, the call `sch = Tables.schema(Tables.columns(first(stream)))` and compare that w/ each partition schema we iterate. Probably a bit simpler than the arrow flatbuffer schema comparing.

Reply:
done in https://github.com/JuliaData/Arrow.jl/pull/160/commits/6a56db3ad923e97a55f7131a86698a230ebcd9bd